felix/fv/qos_controls_test.go

```diff
@@ -289,6 +298,185 @@ var _ = infrastructure.DatastoreDescribe(
+					Expect(err).NotTo(HaveOccurred())
+				}
+
+				By("Removing all limits from workload 1")
+				w[1].WorkloadEndpoint.Spec.QoSControls = nil
+				w[1].UpdateInInfra(infra)
+				Eventually(tc.Felixes[1].ExecOutputFn("ip", "r", "get", "10.65.1.2"), "10s").Should(ContainSubstring(w[1].InterfaceName))
+
+				By("Waiting for the config to disappear in 'iptables-save/nft list ruleset' on workload 1")
+				if NFTMode() {
+					// ingress config should not be present
+					Consistently(getRules(1), "10s", "1s").ShouldNot(MatchRegexp(`(?s)chain filter-cali-tw-` + w[1].InterfaceName + ` {.*\n.*meta l4proto tcp counter packets \d+ bytes \d+ ct count over 4 reject with tcp reset comment ".*Reject connections over ingress connection limit"`))
+					// egress config should not be present
+					Eventually(getRules(1), "10s", "1s").ShouldNot(MatchRegexp(`(?s)chain filter-cali-fw-` + w[1].InterfaceName + ` {.*\n.*meta l4proto tcp counter packets \d+ bytes \d+ ct count over 4 reject with tcp reset comment ".*Reject connections over egress connection limit"`))
+				} else {
+					// ingress config should not be present
+					Consistently(getRules(1), "10s", "1s").ShouldNot(MatchRegexp(`-A cali-tw-` + regexp.QuoteMeta(w[1].InterfaceName) + ` .* -m comment --comment "Reject connections over ingress connection limit" -m conntrack --ctstate NEW,RELATED,ESTABLISHED -m connlimit --connlimit-above ` + fmt.Sprintf("%d", numConnections) + ` --connlimit-mask 0 --connlimit-saddr -j REJECT --reject-with tcp-reset`))
+					// egress config should not be present
+					Eventually(getRules(1), "10s", "1s").ShouldNot(MatchRegexp(`-A cali-fw-` + regexp.QuoteMeta(w[1].InterfaceName) + ` .* -m comment --comment "Reject connections over egress connection limit" -m conntrack --ctstate NEW,RELATED,ESTABLISHED -m connlimit --connlimit-above ` + fmt.Sprintf("%d", numConnections) + ` --connlimit-mask 0 --connlimit-saddr -j REJECT --reject-with tcp-reset`))
+				}
+
+				By("Starting persistent netcat connections on workload 1")
+				for i := range len(netcatCmds) {
+					netcatCmds[i] = w[1].ExecCommand("nc", w[0].IP, "80")
+					err = netcatCmds[i].Start()
+					Expect(err).NotTo(HaveOccurred())
+				}
+
+				By("Running wget (n+1th connection) on workload 1, expecting success")
+				Eventually(tryExecOutput(w[1], "wget", w[0].IP), "10s", "1s").ShouldNot(HaveOccurred())
+				logrus.Infof("%dth connection succeeded as expected", numConnections)
+
+				By("Killing and cleaning up the persistent netcat connection processes")
+				for i := range len(netcatCmds) {
+					err = netcatCmds[i].Process.Kill()
+					Expect(err).NotTo(HaveOccurred())
+					err = netcatCmds[i].Process.Release()
+					Expect(err).NotTo(HaveOccurred())
+				}
+
+				By("Killing and cleaning up nginx server process")
+				err = serverCmd.Process.Kill()
+				Expect(err).NotTo(HaveOccurred())
+				err = serverCmd.Process.Release()
+				Expect(err).NotTo(HaveOccurred())
+			})
+		})
```

Comment:
This is great testing, but I think it would be worth seeing what common functions we can define across the whole of the QoS testing, so that the tests become more easily comprehensible overall.  I've suggested one above, which I think is a really easy one.  Beyond that, I don't think further work is strictly needed for this PR, but I do think that it would help longer term.  The next place I'd look would be the repeating pattern for setting a QoSControl and then waiting for that to be reflected in iptables or TC.